File: logger/src/main/java/org/jetbrains/bsp/bazel/logger/BspClientLogger.java
package org.jetbrains.bsp.bazel.logger;

import com.jetbrains.bsp.bsp4kt.BuildClient;
import com.jetbrains.bsp.bsp4kt.LogMessageParams;
import com.jetbrains.bsp.bsp4kt.MessageType;
import java.time.Duration;
import java.util.function.Supplier;
import org.jetbrains.bsp.bazel.commons.Format;
import org.jetbrains.bsp.bazel.commons.Stopwatch;

public class BspClientLogger {

  private static final Duration LOG_OPERATION_THRESHOLD = Duration.ofMillis(100);
  private BuildClient bspClient;
  private String originId;

  public BspClientLogger withOriginId(String originId) {
    BspClientLogger bspClientLogger = new BspClientLogger();
    bspClientLogger.originId = originId;
    bspClientLogger.bspClient = bspClient;
    return bspClientLogger;
  }

  public void error(String errorMessage) {
    log(MessageType.Error, errorMessage);
  }

  public void message(String format, Object... args) {
    message(String.format(format, args));
  }

  public void message(String message) {
    log(MessageType.Log, message);
  }

  public void warn(String format, Object... args) {
    warn(String.format(format, args));
  }

  public void warn(String message) {
    log(MessageType.Warning, message);
  }

  private void log(MessageType messageType, String message) {
    if (bspClient == null) return;

    if (!message.trim().isEmpty()) {
      var params = new LogMessageParams(messageType, null, originId, message);
      bspClient.onBuildLogMessage(params);
    }
  }

  public <T> T timed(String description, Supplier<T> supplier) {
    var sw = Stopwatch.start();
    T result = supplier.get();
    var duration = sw.stop();
    logDuration(description, duration);
    return result;
  }

  public void logDuration(String description, Duration duration) {
    if (duration.compareTo(LOG_OPERATION_THRESHOLD) >= 0) {
      message("Task '%s' completed in %s", description, Format.duration(duration));
    }
  }

  public void timed(String description, Runnable runnable) {
    timed(
        description,
        () -> {
          runnable.run();
          return null;
        });
  }

  public void initialize(BuildClient buildClient) {
    this.bspClient = buildClient;
  }
}

File: logger/src/main/kotlin/org/jetbrains/bsp/bazel/logger/BspClientLogger.kt
package org.jetbrains.bsp.bazel.logger

import ch.epfl.scala.bsp4j.BuildClient
import ch.epfl.scala.bsp4j.LogMessageParams
import ch.epfl.scala.bsp4j.MessageType
import org.jetbrains.bsp.bazel.commons.Format
import java.time.Duration

private val LOG_OPERATION_THRESHOLD: Duration = Duration.ofMillis(100)

data class BspClientLogger(private val bspClient: BuildClient, private val originId: String? = null) {

  fun error(errorMessage: String) {
    log(MessageType.ERROR, errorMessage)
  }

  fun message(format: String, vararg args: Any?) {
    message(String.format(format, *args))
  }

  fun message(message: String) {
    log(MessageType.LOG, message)
  }

  fun warn(format: String, vararg args: Any?) {
    warn(String.format(format, *args))
  }

  fun warn(message: String) {
    log(MessageType.WARNING, message)
  }

  private fun log(messageType: MessageType, message: String) {
    if (message.trim { it <= ' ' }.isNotEmpty()) {
      val params = LogMessageParams(messageType, message)
      params.originId = originId
      bspClient.onBuildLogMessage(params)
    }
  }

  fun logDuration(description: String, duration: Duration) {
    if (duration >= LOG_OPERATION_THRESHOLD) {
      message("Task '%s' completed in %s", description, Format.duration(duration))
    }
  }
}
